### PR TITLE
enhance container export api

### PR DESCRIPTION
**Pull request by EvanLjp**
# Description
The feature is for exporting the container meta to the CPP part. And I found the CRI `loopSyncContainers` method has some bug which is the k8s info would not be appended to the real container until the `fetchAll`  method is triggered.

# Changes
- [public] [both] [fixed] fixed without k8s info in sync containers when using CRI runtime 
- [public] [both] [updated] force fetch container info when not found in the cache for export GetContainerMeta API  

# Design
1. ContainerDiscoverManager would control all fetching container meta operations, it would start some go routines to do it.
2. Origin logic.
   - start a goroutine to fetch all for docker,CRI, and static.
   - start a goroutine to listen to docker events when enabling docker fetching.
   - start a goroutine to sync CRI containers when enabling CRI fetching.
   - start a goroutine to read static info when enabling static fetching.
3. New logic.
   - export Fetchone method to inspect container ID.

### CODE REVIEW

File: main/plugin_export.go [hunk 1]
```diff
@@ -29,11 +29,26 @@ import (
...
+static void setArrayString(char **a, char *s, int n) {
+        a[n] = s;
+}
+
 struct containerMeta{
 	char* podName;
 	char* k8sNamespace;
 	char* containerName;
 	char* image;
+	int k8sLabelsSize;
```
Comment (shabicheng) [inline main/plugin_export.go:46]: 如果要加的话，可以把环境变量也加上

File: helper/container_finding_controller.go [hunk 2]
```diff
@@ -0,0 +1,269 @@
+// Copyright 2021 iLogtail Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (messixukejia) [inline helper/container_finding_controller.go:3]: 代码层次改下：
总的控制文件（与容器运行时无关）-> container_finding_controller -> docker_center/docker_cri_adaptor

File: helper/docker_center.go [hunk 3]
```diff
@@ -1130,145 +1113,78 @@ func dockerCenterRecover() {
...
-			time.Sleep(time.Duration(10) * time.Second)
-			logger.Debug(context.Background(), "docker clean timeout container", "start")
-			dc.cleanTimeoutContainer()
-			logger.Debug(context.Background(), "docker clean timeout container", "done")
-			if time.Since(lastFetchAllTime) >= FetchAllInterval {
-				logger.Info(context.Background(), "docker fetch all", "start")
-				dc.readStaticConfig(true)
-				fetchErr := dc.fetchAll()
-				if fetchErr != nil {
-					fetchErrCount++
```
Comment (messixukejia) [inline helper/docker_center.go:1192]: 这个逻辑干掉了？

File: helper/container_finding_controller.go [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+
+func (c *ContainerDiscoverManager) fetchCRI() error {
+	if criRuntimeWrapper == nil {
+		return nil
+	}
+	return criRuntimeWrapper.fetchAll()
+}
+
+func (c *ContainerDiscoverManager) SyncContainers() {
+	if c.enableCRIDiscover {
```
Comment (messixukejia) [inline helper/container_finding_controller.go:126]: if 判断很多上下文不一致，一会用enableCRIDiscover、一会用criRuntimeWrapper

File: main/plugin_export.go [hunk 5]
```diff
@@ -150,10 +169,74 @@ func GetContainerMeta(containerID string) *C.struct_containerMeta {
...
+		return convertFunc(detail)
+	}
+
+	if manager := helper.GetContainerFindingManager(); manager != nil {
+		if err := manager.FetchOne(containerID); err != nil {
+			logger.Debugf(context.Background(), "cannot fetch container for %s, error is %v", containerID, err)
+			return nil
+		}
+	}
+	detail, ok = helper.GetDockerCenterInstance().GetContainerDetail(containerID)
```
Comment (messixukejia) [inline main/plugin_export.go:236]: 这里也没有去获取sandbox的信息，为什么FetchOne中去获取了一次，且存入updateContainer？

File: helper/docker_center.go [hunk 6]
```diff
@@ -1061,14 +1033,25 @@ func (dc *DockerCenter) fetchAll() error {
...
 		dc.updateContainer(containerID, dc.CreateInfoDetail(containerDetail, envConfigPrefix, false))
 	}
 	logger.Debug(context.Background(), "update container", containerID, "detail", containerDetail)
+	if tryFindSandbox && containerDetail.Config != nil {
+		if id := containerDetail.Config.Labels["io.kubernetes.sandbox.id"]; id != "" {
+			containerDetail, err = dc.client.InspectContainerWithOptions(docker.InspectContainerOptions{ID: id})
+			if err != nil {
+				dc.setLastError(err, "inspect sandbox container error "+id)
+			} else {
+				dc.updateContainer(id, dc.CreateInfoDetail(containerDetail, envConfigPrefix, false))
```
Comment (messixukejia) [inline helper/docker_center.go:1050]: 这里存sandbox目的是什么？fetchall的时候却不获取也不存？

File: helper/container_finding_controller.go [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+		logger.Debug(context.Background(), "discover manager start sync containers goroutine", "cri")
+		go criRuntimeWrapper.loopSyncContainers()
+	}
+	if c.enableStaticDiscover {
+		logger.Debug(context.Background(), "discover manager start sync containers goroutine", "static")
```
Comment (messixukejia) [inline helper/container_finding_controller.go:131]: 原有逻辑是先dockerCenterInstance.readStaticConfig(true)，为什么去掉了？

File: helper/container_finding_controller.go [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+	timerFetch := func() {
+		defer dockerCenterRecover()
+		lastFetchAllTime := time.Now()
+		for {
+			time.Sleep(time.Duration(10) * time.Second)
+			logger.Debug(context.Background(), "container clean timeout container info", "start")
+			dockerCenterInstance.cleanTimeoutContainer()
+			logger.Debug(context.Background(), "container clean timeout container info", "done")
+			if time.Since(lastFetchAllTime) >= FetchAllInterval {
+				logger.Info(context.Background(), "container fetch all", "start")
```
Comment (messixukejia) [inline helper/container_finding_controller.go:259]: 这里去掉了readStaticConfig？之前不管cri还是docker都是在for循环里一直掉用的

File: helper/docker_center.go (hunk 6 above)
Comment (EvanLjp) [reply helper/docker_center.go:1050]: 之前介绍过了，docker 监听机制，会有sandbox 触发，getContainerByID 只有容器ID 所以需要自行补

File: helper/container_finding_controller.go (hunk 8 above)
Comment (EvanLjp) [reply helper/container_finding_controller.go:259]: fetchALL 第一行就会调用

File: helper/container_finding_controller.go (hunk 4 above)
Comment (EvanLjp) [reply helper/container_finding_controller.go:126]: 这个我看一下

File: helper/container_finding_controller.go (hunk 7 above)
Comment (EvanLjp) [reply helper/container_finding_controller.go:131]: 在 init里

File: helper/docker_center.go [hunk 9]
```diff
@@ -632,27 +606,22 @@ func GetDockerCenterInstance() *DockerCenter {
...
-		if isStaticContainerInfoEnabled() {
-			dockerCenterInstance.readStaticConfig(true)
-			go dockerCenterInstance.flushStaticConfig()
-		}
-
+		var enableCriFinding = criRuntimeWrapper != nil
+		var enableDocker = dockerCenterInstance.InitClient() == nil
+		var enableStatic = isStaticContainerInfoEnabled()
+		containerFindingManager = NewContainerDiscoverManager(enableDocker, enableCriFinding, enableStatic)
+		containerFindingManager.Init(3)
```
[the remark this answers is not in the document]
Comment (EvanLjp) [reply helper/docker_center.go:613]: 原来是几次？麻烦贴一下，没找到相关内容

**Pull request merged**